x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java

```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
         return INTERNAL_PREDICATE.test(action);
     }
+
+    public static boolean isParentAction(String action) {
+        return PRE_AUTHORIZED_CHILD_ACTIONS_BY_PARENT.containsKey(action);
+    }
+
+    public static boolean shouldPreAuthorizeChildAction(final String parent, final String child) {
+        final Set<String> children = PRE_AUTHORIZED_CHILD_ACTIONS_BY_PARENT.get(parent);
+        return children != null && (parent.equals(child) || children.contains(child));
+    }
+
+    public static void putParentAuthorizationContext(ThreadContext threadContext, AuthorizationEngine.AuthorizationContext context) {
+        AuthorizationEngine.AuthorizationContext existing = threadContext.getTransient(
+            AuthorizationServiceField.PARENT_AUTHORIZATION_CONTEXT_KEY
+        );
+        if (existing == null) {
+            threadContext.putTransient(AuthorizationServiceField.PARENT_AUTHORIZATION_CONTEXT_KEY, context);
+        } else if (existing.getAction().equals(context.getAction()) == false) {
+            throw new AssertionError("parent authorization context already set in thread context");
+        }
+    }
+
+    public static AuthorizationEngine.AuthorizationContext readParentAuthorizationContext(ThreadContext threadContext) {
+        return threadContext.getTransient(AuthorizationServiceField.PARENT_AUTHORIZATION_CONTEXT_KEY);
+    }
+
+    public static void maybePreAuthorizeChildAction(
+        ThreadContext threadContext,
+        Transport.Connection connection,
+        TransportRequest request,
+        String childAction,
+        Version version,
+        ClusterState clusterState
+
+    ) {
+        if (RemoteConnectionManager.resolveRemoteClusterAlias(connection).isPresent()) {
+            // We can only pre-authorize actions targeting a node which belongs to the same cluster.
+            if (threadContext.getHeader(ParentActionAuthorization.THREAD_CONTEXT_KEY) != null) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(
+                        "attempted to send the request ["
+                            + childAction
+                            + "] to the remote cluster ["
+                            + RemoteConnectionManager.resolveRemoteClusterAlias(connection).get()
+                            + "] with the pre-authorization header ["
+                            + threadContext.getHeader(ParentActionAuthorization.THREAD_CONTEXT_KEY)
+                            + "]"
+                    );
+                }
+                throw new AssertionError("pre-authorization not expected to be set for remote cluster requests");
+            }
+            return;
+        }
+
+        final AuthorizationEngine.AuthorizationContext parentContext = readParentAuthorizationContext(threadContext);
+        if (parentContext == null) {
+            return;
+        }
+
+        final IndicesAccessControl indicesAccessControl = parentContext.getIndicesAccessControl();
```

Comment:
You've checked for DLS/FLS in other places, but I think this is the best place do it: i.e. when deciding to pre-authorize.

Reply:
I've checked on both receiving (remote node) and sending side (coordinating node).
Reason is to make sure DLS/FLS is not enabled on all nodes. If I were to check it only here (on coordinating node) it is possible that roles are configured differently on remote nodes and we would potentially pre-authorize child action without respecting DLS/FLS.

Reply:
I agree, checking that the role doesn't have DLS/FLS on both the parent action (just before setting the header), and on the child action (just before "bypassing" regular authorization sound like the best and cautious option.